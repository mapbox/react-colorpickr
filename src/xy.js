'use strict';

var React = require('react');
var ReactDOM = require('react-dom');
var clamp = require('clamp');

var XYControl = React.createClass({
  propTypes: {
    x: React.PropTypes.number.isRequired,
    y: React.PropTypes.number.isRequired,
    xmax: React.PropTypes.number.isRequired,
    ymax: React.PropTypes.number.isRequired,
    handleClass: React.PropTypes.string,
    onChange: React.PropTypes.func.isRequired
  },
  getInitialState() {
    return {
      dragging: false
    };
  },
  change(pos) {
    var rect = this.getOwnBoundingRect();
    this.props.onChange({
      x: clamp(pos.left, 0, rect.width) / rect.width * this.props.xmax,
      y: clamp(pos.top, 0, rect.height) / rect.height * this.props.ymax
    });
  },
  getOwnBoundingRect() {
    return ReactDOM.findDOMNode(this).getBoundingClientRect();
  },
  _onMouseDown(e) {
    var rect = this.getOwnBoundingRect();
    var x = e.clientX,
      y = e.clientY;

    var offset = {
      left: x - rect.left,
      top: y - rect.top
    };

    this.change(offset);

    // Handle interaction
    this.setState({
      start: { x: offset.left, y: offset.top },
      offset: { x, y }
    });

    window.addEventListener('mousemove', this._drag);
    window.addEventListener('mouseup', this._dragEnd);
  },
  _drag(e) {
    this.setState({ dragging: true });
    var posX = e.clientX + this.state.start.x - this.state.offset.x;
    var posY = e.clientY + this.state.start.y - this.state.offset.y;

    this.change({
      left: posX,
      top: posY
    });
  },
  _dragEnd() {
    this.setState({ dragging: false });
    window.removeEventListener('mousemove', this._drag);
    window.removeEventListener('mouseup', this._dragEnd);
  },
  render() {
    return (
      <div
        onMouseDown={this._onMouseDown}
        className={`slider-xy ${this.state.dragging ? 'dragging-xy' : ''}`}>
        <div
          className={`handle-xy ${this.props.handleClass}`}
          style={{
            top: clamp(this.props.y / this.props.ymax * 100, 0, 100) + '%',
            left: clamp(this.props.x / this.props.xmax * 100, 0, 100) + '%'
          }} />
      </div>
    );
  }
});

module.exports = XYControl;
