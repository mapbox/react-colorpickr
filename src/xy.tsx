import React, { useState, useRef } from 'react';
import PropTypes from 'prop-types';
import themeable from 'react-themeable';
import { autokey } from './autokey.ts';
import clamp from 'clamp';

function XYControl({ children, theme, x, y, xmax, ymax, isDark, onChange }) {
  const xyControlContainer = useRef(null);
  const [coords, setCoords] = useState({ start: {}, offset: {}, cb: null });
  const [active, setActive] = useState(false);
  const top = Math.round(clamp((y / ymax) * 100, 0, 100));
  const left = Math.round(clamp((x / xmax) * 100, 0, 100));
  const themer = autokey(themeable(theme));

  const change = (pos) => {
    const rect = xyControlContainer.current.getBoundingClientRect();
    onChange({
      x: (clamp(pos.left, 0, rect.width) / rect.width) * xmax,
      y: (clamp(pos.top, 0, rect.height) / rect.height) * ymax
    });
  };

  const dragStart = (e) => {
    e.preventDefault();
    setActive(true);

    const rect = xyControlContainer.current.getBoundingClientRect();
    const x = e.changedTouches ? e.changedTouches[0].clientX : e.clientX;
    const y = e.changedTouches ? e.changedTouches[0].clientY : e.clientY;

    const offset = {
      left: x - rect.left,
      top: y - rect.top
    };

    change(offset);

    setCoords({
      start: { x: offset.left, y: offset.top },
      offset: { x, y }
    });
  };

  const drag = (e) => {
    if (!active) return;
    const { start, offset } = coords;
    e.preventDefault();
    const top =
      (e.changedTouches ? e.changedTouches[0].clientY : e.clientY) +
      start.y -
      offset.y;
    const left =
      (e.changedTouches ? e.changedTouches[0].clientX : e.clientX) +
      start.x -
      offset.x;

    change({ top, left });
  };

  const dragEnd = () => {
    setActive(false);
  };

  return (
    <div
      {...themer('xyControlContainer')}
      data-testid="xy"
      ref={xyControlContainer}
      onTouchStart={dragStart}
      onTouchMove={drag}
      onTouchEnd={dragEnd}
      onMouseDown={dragStart}
      onMouseMove={drag}
      onMouseUp={dragEnd}
    >
      <div
        {...themer('xyControl', `${isDark ? 'xyControlDark' : ''}`)}
        style={{
          top: `${top}%`,
          left: `${left}%`
        }}
      />
      {children}
    </div>
  );
}

XYControl.propTypes = {
  children: PropTypes.node.isRequired,
  theme: PropTypes.object.isRequired,
  x: PropTypes.number.isRequired,
  y: PropTypes.number.isRequired,
  xmax: PropTypes.number.isRequired,
  ymax: PropTypes.number.isRequired,
  isDark: PropTypes.bool.isRequired,
  onChange: PropTypes.func.isRequired
};

export { XYControl };
