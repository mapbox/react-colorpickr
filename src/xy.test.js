import React from 'react';
import { render } from '@testing-library/react';
import { XYControl } from './xy';

describe('XYControl', () => {
  test('renders', () => {
    const props = {
      theme: {},
      x: 0,
      y: 10,
      xmax: 100,
      ymax: 100,
      isDark: false,
      onChange: jest.fn()
    };

    const { baseElement } = render(
      <XYControl {...props}>
        <span>children</span>
      </XYControl>
    );
    expect(baseElement).toMatchSnapshot();
  });

  test('renders isDark', () => {
    const props = {
      theme: {},
      x: 0,
      y: 10,
      xmax: 100,
      ymax: 100,
      isDark: true,
      onChange: jest.fn()
    };

    const { baseElement } = render(
      <XYControl {...props}>
        <span>children</span>
      </XYControl>
    );
    expect(baseElement).toMatchSnapshot();
  });
});
