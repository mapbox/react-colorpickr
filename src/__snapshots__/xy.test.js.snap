// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`XYControl renders 1`] = `
<body>
  <div>
    <div>
      <div
        style="top: 10%; left: 0%;"
      />
      <span>
        children
      </span>
    </div>
  </div>
</body>
`;

exports[`XYControl renders isDark 1`] = `
<body>
  <div>
    <div>
      <div
        style="top: 10%; left: 0%;"
      />
      <span>
        children
      </span>
    </div>
  </div>
</body>
`;
